In [1]:
# Import libraries
import pandas as pd
import numpy as np
import os


In [2]:
path = r'C:\Users\honoc\Documents\CareerFoundry Folder\Python Fundamentals for Data Analysts\Exercise 4.3 Folder'
path
df_ords_prods_merged = pd.read_pickle(os.path.join(path, 'Prepared data', 'orders_products_merged.pkl'))

In [3]:
df_ords_prods_merged.head()

Unnamed: 0  product_id                product_name  aisle_id  \
0           0           1  Chocolate Sandwich Cookies        61   
1           0           1  Chocolate Sandwich Cookies        61   
2           0           1  Chocolate Sandwich Cookies        61   
3           0           1  Chocolate Sandwich Cookies        61   
4           0           1  Chocolate Sandwich Cookies        61   

   department_id  prices  order_id  user_id  order_number  orders_day_of_week  \
0             19     5.8   3139998      138            28                   6   
1             19     5.8   1977647      138            30                   6   
2             19     5.8    389851      709             2                   0   
3             19     5.8   1813452      764             3                   4   
4             19     5.8   1701441      777            16                   1   

   order_hour_of_day  days_since_prior_order  add_to_cart_order  reordered  \
0                 11                     3.0                  5          0   
1                 17                    20.0                  1          1   
2                 21                     6.0                 20          0   
3                 17                     9.0                 11          1   
4                  7                    26.0                  7          0   

  _merge  
0   both  
1   both  
2   both  
3   both  
4   both

In [4]:
# Dropping columns = ["Unnamed", "_merge"] from dataframe
columns_to_remove = ["Unnamed: 0", "_merge"]
df_ords_prods_merged = df_ords_prods_merged.drop(columns= columns_to_remove)

In [5]:
df_ords_prods_merged

product_id                product_name  aisle_id  department_id  \
0                  1  Chocolate Sandwich Cookies        61             19   
1                  1  Chocolate Sandwich Cookies        61             19   
2                  1  Chocolate Sandwich Cookies        61             19   
3                  1  Chocolate Sandwich Cookies        61             19   
4                  1  Chocolate Sandwich Cookies        61             19   
...              ...                         ...       ...            ...   
30328758       49688      Fresh Foaming Cleanser        73             11   
30328759       49688      Fresh Foaming Cleanser        73             11   
30328760       49688      Fresh Foaming Cleanser        73             11   
30328761       49688      Fresh Foaming Cleanser        73             11   
30328762       49688      Fresh Foaming Cleanser        73             11   

          prices  order_id  user_id  order_number  orders_day_of_week  \
0            5.8   3139998      138            28                   6   
1            5.8   1977647      138            30                   6   
2            5.8    389851      709             2                   0   
3            5.8   1813452      764             3                   4   
4            5.8   1701441      777            16                   1   
...          ...       ...      ...           ...                 ...   
30328758    13.5   3106363   197371            10                   4   
30328759    13.5   1788356   200215             2                   0   
30328760    13.5    809510   200873             5                   3   
30328761    13.5   2359893   200873             9                   3   
30328762    13.5   2385091   205926            11                   1   

          order_hour_of_day  days_since_prior_order  add_to_cart_order  \
0                        11                     3.0                  5   
1                        17                    20.0                  1   
2                        21                     6.0                 20   
3                        17                     9.0                 11   
4                         7                    26.0                  7   
...                     ...                     ...                ...   
30328758                 14                    30.0                 17   
30328759                  9                     5.0                 27   
30328760                  8                    15.0                 12   
30328761                 15                     5.0                 11   
30328762                 15                     6.0                 35   

          reordered  
0                 0  
1                 1  
2                 0  
3                 1  
4                 0  
...             ...  
30328758          0  
30328759          0  
30328760          0  
30328761          1  
30328762          0  

[30328763 rows x 13 columns]

In [6]:
df_ords_prods_merged.shape

(30328763, 13)

# Creating a subset

In [7]:
df = df_ords_prods_merged[:1000000]

# Capturing the below "If" statements in codes

If the value within the “prices” column within the given row is less than or equal to 5, then return the string “Low-range product.”

Or else, if the value within the “prices” column within the given row is greater than 5 and less than or equal to 15, then return the string “Mid-range product.”

Or else, if the value of the “prices” column within the given row is greater than 15, then return the string “High-range product.”

Or else, return the string “Not enough data.”

In [8]:
df.shape

(1000000, 13)

In [9]:
# Define a function that captures the above "If" statements
def price_label(row):
    
    if row['prices'] <= 5:
        return 'Low range product'
    elif (row['prices'] > 5) and (row['prices'] <= 15):
        return 'Mid range product'
    elif row['prices'] > 15:
        return 'High range product'
    else: return np.nan

# Creating a new column called "price_range" within the dataframe "df". the "price_range" column would be designated as a location for the labels.

In [10]:
# Apply the function
df['price_range'] = df.apply(price_label, axis=1) # axis=1 starts with the rows and not the columns

C:\Users\honoc\AppData\Local\Temp\ipykernel_52292\2238658058.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['price_range'] = df.apply(price_label, axis=1) # axis=1 starts with the rows and not the columns


In [11]:
df['price_range'].value_counts(dropna= False)

price_range
Mid range product     653498
Low range product     337952
High range product      8550
Name: count, dtype: int64

In [12]:
# Checking for the most expensive product within the subset with the max() function
df['prices'].max()

24.5

C:\Users\honoc\AppData\Local\Temp\ipykernel_36740\2238658058.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

If-Statements with the loc() Function which can apply the conditional logic of an if-statement to a function without explicitly creating an if-else construct.

In [13]:
df.loc[df['prices'] > 15, 'price_range_loc'] = 'High range product'

C:\Users\honoc\AppData\Local\Temp\ipykernel_52292\1580616944.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df.loc[df['prices'] > 15, 'price_range_loc'] = 'High range product'


In [14]:
df.loc[(df['prices'] <= 15) & (df['prices'] > 5), 'price_range_loc'] = 'Mid range product'

In [15]:
df.loc[df['prices'] <= 5, 'price_range_loc'] = 'Low range product'

In [16]:
df['price_range_loc'].value_counts(dropna= False)

price_range_loc
Mid range product     653498
Low range product     337952
High range product      8550
Name: count, dtype: int64

# Using the loc() function on entire dataframe

In [17]:
df_ords_prods_merged.loc[df_ords_prods_merged['prices'] > 15, 'price_range_loc'] = 'High-range product'

In [18]:
df_ords_prods_merged.loc[(df_ords_prods_merged['prices'] <= 15) & (df_ords_prods_merged['prices'] > 5), 'price_range_loc'] = 'Mid-range product'

In [19]:
df_ords_prods_merged.loc[df_ords_prods_merged['prices'] <= 5, 'price_range_loc'] = 'Low-range product'

In [20]:
df_ords_prods_merged['price_range_loc'].value_counts(dropna= False)

price_range_loc
Mid-range product     20462144
Low-range product      9476774
High-range product      389845
Name: count, dtype: int64

# If-Statements with For-Loops. Application: To get an insight into what products are being bought on the busiest and slowest days. As information could be used to tailor ads on specific days. 

In [21]:
# first we need to know on which day most orders take place. This is achieved by printing the frequency of the "orders_day_of_the_week" column.
df_ords_prods_merged['orders_day_of_week'].value_counts(dropna= False)

orders_day_of_week
0    5779087
1    5303718
6    4190948
5    3952326
2    3947564
3    3600589
4    3554531
Name: count, dtype: int64

# Creating a column "busiest day" that will contain one of three different values: "Busiest day," "Least busy," and "Regularly busy"

In [22]:
result = []

for value in df_ords_prods_merged["orders_day_of_week"]:
    if value == 0:
        result.append("Busiest day")
    elif value == 4:
        result.append("Least busy")
    else:
        result.append("Regularly busy")

In [23]:
result

['Regularly busy',
 'Regularly busy',
 'Busiest day',
 'Least busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Least busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Busiest day',
 'Busiest day',
 'Busiest day',
 'Busiest day',
 'Busiest day',
 'Least busy',
 'Regularly busy',
 'Regularly busy',
 'Busiest day',
 'Regularly busy',
 'Least busy',
 'Busiest day',
 'Regularly busy',
 'Busiest day',
 'Least busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Busiest day',
 'Regularly busy',
 'Busiest day',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regula

# Combining this result with the "df_ords_prods_merged" dataframe by creating a new column "busiest_day" and setting it equal to "result"

In [24]:
df_ords_prods_merged['busiest_day'] = result
df_ords_prods_merged['busiest_day'].value_counts(dropna= False)

busiest_day
Regularly busy    20995145
Busiest day        5779087
Least busy         3554531
Name: count, dtype: int64

2.	Suppose your clients have changed their minds about the labels you created in your “busiest_day” column. Now, they want “Busiest day” to become “Busiest days” (plural). This label should correspond with the two busiest days of the week as opposed to the single busiest day. At the same time, they’d also like to know the two slowest days. Create a new column for this using a suitable method.

In [25]:
# Creating a list to store the labels for the two busiest and two slowest days
busiest_and_slowest_days = []

In [26]:
# Define the busiest and slowest days
busiest_days = [0, 1] # Saturday and Sunday
slowest_days = [3, 4]

In [27]:
# Update the "busiest_day" column and create a new column "slowest_day"
for value in df_ords_prods_merged['orders_day_of_week']:
    if value in busiest_days:
        busiest_and_slowest_days.append("Busiest days")
    elif value in slowest_days:
        busiest_and_slowest_days.append("Slowest days")
    else:
        busiest_and_slowest_days.append("Regularly busy")

3.	Check the values of this new column for accuracy. Note any observations in markdown format.

In [29]:
df_ords_prods_merged['busiest_day'] = busiest_and_slowest_days
df_ords_prods_merged['busiest_day'].value_counts(dropna= False)

busiest_day
Regularly busy    12090838
Busiest days      11082805
Slowest days       7155120
Name: count, dtype: int64

"Regularly busy" appears most frequently, indicating that most days fall into this category. "Busiest days" have the next highest count, representing the two busiest days, i.e., Saturday and Sunday. "Slowest days" have the lowest count, signifying the two slowest days, i.e., Tuesday and Wednesday. The total count  = 30,328,763 same as df_ords_prods_merged.shape (30328763, 13)

In [32]:
# Add the new columns to the Dataframe
df_ords_prods_merged['busiest_day'] = busiest_and_slowest_days
df_ords_prods_merged['slowest_day'] = busiest_and_slowest_days

In [33]:
df_ords_prods_merged.head()

product_id                product_name  aisle_id  department_id  prices  \
0           1  Chocolate Sandwich Cookies        61             19     5.8   
1           1  Chocolate Sandwich Cookies        61             19     5.8   
2           1  Chocolate Sandwich Cookies        61             19     5.8   
3           1  Chocolate Sandwich Cookies        61             19     5.8   
4           1  Chocolate Sandwich Cookies        61             19     5.8   

   order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   3139998      138            28                   6                 11   
1   1977647      138            30                   6                 17   
2    389851      709             2                   0                 21   
3   1813452      764             3                   4                 17   
4   1701441      777            16                   1                  7   

   days_since_prior_order  add_to_cart_order  reordered    price_range_loc  \
0                     3.0                  5          0  Mid-range product   
1                    20.0                  1          1  Mid-range product   
2                     6.0                 20          0  Mid-range product   
3                     9.0                 11          1  Mid-range product   
4                    26.0                  7          0  Mid-range product   

      busiest_day     slowest_day  
0  Regularly busy  Regularly busy  
1  Regularly busy  Regularly busy  
2    Busiest days    Busiest days  
3    Slowest days    Slowest days  
4    Busiest days    Busiest days

4. When too many users make Instacart orders at the same time, the app freezes. The senior technical officer at Instacart wants you to identify the busiest hours of the day. Rather than by hour, they want periods of time labeled “Most orders,” “Average orders,” and “Fewest orders.” Create a new column containing these labels called “busiest_period_of_day.”

In [36]:
# first we need to know on which day most orders take place. This is achieved by printing the frequency of the "orders_day_of_the_week" column.
df_ords_prods_merged['order_hour_of_day'].value_counts(dropna= False)

order_hour_of_day
10    2593725
11    2564597
14    2517238
15    2487586
13    2487500
12    2445841
16    2364969
9     2311334
17    1943858
8     1622394
18    1520954
19    1169224
20     910005
7      844665
21     746254
22     592432
23     375889
6      274801
0      203460
1      108110
5       82706
2       63961
4       49400
3       47860
Name: count, dtype: int64

In [37]:
# Create a list to store the labels for busiest periods of the day
busiest_period_of_day = []

In [38]:
# Define the criteria for busiest, average, and fewest orders
most_orders_hours = [10, 11, 14, 15, 13, 12] # Hours with the highest orders
average_orders_hours = [16, 9, 17, 8, 18, 19] # Hours with moderate order counts
fewest_orders_hours = [20, 7, 21, 22, 23, 6, 0, 1, 5, 2, 4, 3] # Hours with the lowest order counts

In [39]:
# Update the "busiest_period_of_day" column
for hour in df_ords_prods_merged['order_hour_of_day']:
    if hour in most_orders_hours:
        busiest_period_of_day.append("Most orders")
    elif hour in average_orders_hours:
        busiest_period_of_day.append("Average orders")
    else: 
        busiest_period_of_day.append("Fewest orders")

In [40]:
# Add the new "busiest_period_of_day" column to the Dataframe
df_ords_prods_merged['busiest_period_of_day'] = busiest_period_of_day

In [41]:
df_ords_prods_merged.head()

product_id                product_name  aisle_id  department_id  prices  \
0           1  Chocolate Sandwich Cookies        61             19     5.8   
1           1  Chocolate Sandwich Cookies        61             19     5.8   
2           1  Chocolate Sandwich Cookies        61             19     5.8   
3           1  Chocolate Sandwich Cookies        61             19     5.8   
4           1  Chocolate Sandwich Cookies        61             19     5.8   

   order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   3139998      138            28                   6                 11   
1   1977647      138            30                   6                 17   
2    389851      709             2                   0                 21   
3   1813452      764             3                   4                 17   
4   1701441      777            16                   1                  7   

   days_since_prior_order  add_to_cart_order  reordered    price_range_loc  \
0                     3.0                  5          0  Mid-range product   
1                    20.0                  1          1  Mid-range product   
2                     6.0                 20          0  Mid-range product   
3                     9.0                 11          1  Mid-range product   
4                    26.0                  7          0  Mid-range product   

      busiest_day     slowest_day busiest_period_of_day  
0  Regularly busy  Regularly busy           Most orders  
1  Regularly busy  Regularly busy        Average orders  
2    Busiest days    Busiest days         Fewest orders  
3    Slowest days    Slowest days        Average orders  
4    Busiest days    Busiest days         Fewest orders

5.	Print the frequency for this new column.

In [42]:
df_ords_prods_merged['busiest_period_of_day'].value_counts(dropna= False)

busiest_period_of_day
Most orders       15096487
Average orders    10932733
Fewest orders      4299543
Name: count, dtype: int64

In [43]:
df_ords_prods_merged.to_pickle(os.path.join(path, 'Prepared data', 'orders_products_new_variables.pkl'))